### PR TITLE
Laboratory work #4, Ivan Shmakov - 23FPL1

### PR DESCRIPTION
**Pull request by CloudWhisp**

### CODE REVIEW

File: lab_4_retrieval_w_clustering/main.py [hunk 1]
```diff
@@ -4,6 +4,8 @@
 Vector search with clusterization
 """
 
+from lab_2_retrieval_w_bm25.main import calculate_bm25
+
 # pylint: disable=undefined-variable, too-few-public-methods, unused-argument, duplicate-code, unused-private-member, super-init-not-called
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py:7]: ignores before imports

File: lab_4_retrieval_w_clustering/main.py [hunk 2]
```diff
@@ -51,6 +56,11 @@ def set_tokenized_corpus(self, tokenized_corpus: TokenizedCorpus) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not isinstance(tokenized_corpus, list) or len(tokenized_corpus) == 0:
+            raise ValueError
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: add message

File: lab_4_retrieval_w_clustering/main.py [hunk 3]
```diff
@@ -67,6 +77,12 @@ def vectorize(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if not isinstance(tokenized_document, list) or len(tokenized_document) == 0:
+            raise ValueError
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: .

File: lab_4_retrieval_w_clustering/main.py [hunk 4]
```diff
@@ -67,6 +77,12 @@ def vectorize(self, tokenized_document: list[str]) -> Vector:
...
+        vector = self._calculate_bm25(tokenized_document)
+        if vector is None:
+            raise ValueError
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: .

File: lab_4_retrieval_w_clustering/main.py [hunk 5]
```diff
@@ -81,6 +97,20 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if not isinstance(tokenized_document, list) or len(tokenized_document) == 0:
+            raise ValueError
+        vector = [0.0 for elem in range(len(self._vocabulary))]
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: try to use another way to get a list of a certain length with the same values, remember the seminar

File: lab_4_retrieval_w_clustering/main.py [hunk 6]
```diff
@@ -81,6 +97,20 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
...
+        bm = calculate_bm25(self._vocabulary, tokenized_document,
+                            self._idf_values, 1.5, 0.75,
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: use default arguments

File: lab_4_retrieval_w_clustering/main.py [hunk 7]
```diff
@@ -81,6 +97,20 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
...
+                            self._avg_doc_len, len(tokenized_document))
+        if len(self._vocabulary) == 0:
+            return tuple(vector)
+        if bm is None:
+            raise ValueError
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: .

File: lab_4_retrieval_w_clustering/main.py [hunk 8]
```diff
@@ -113,6 +147,28 @@ def put_corpus(self, corpus: Corpus) -> None:
                 or if input arguments are empty,
                 or if methods used return None.
         """
+        if not isinstance(corpus, list) or len(corpus) == 0:
+            raise ValueError
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: .

File: lab_4_retrieval_w_clustering/main.py [hunk 9]
```diff
@@ -182,7 +256,28 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
...
+        tokenized_query = self._tokenizer.tokenize(query)
+        if tokenized_query is None or len(tokenized_query) == 0:
+            raise ValueError
+        query_vector = self._vectorizer.vectorize(tokenized_query)
+        if query_vector is None or len(query_vector) == 0:
+            raise ValueError
+        vector_data = [x[1] for x in self._db.get_vectors()]
+        answer = self._calculate_knn(query_vector, vector_data, n_neighbours)
+        print(answer)
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: rm

File: lab_4_retrieval_w_clustering/main.py [hunk 10]
```diff
@@ -182,7 +256,28 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
...
+        if answer is None or len(answer) == 0:
+            raise ValueError
+        ind = tuple([doc_data[0] for doc_data in answer])
+        print(ind)
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: .

File: lab_4_retrieval_w_clustering/main.py [hunk 11]
```diff
@@ -182,7 +256,28 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
         Returns:
             list[tuple[float, str]]: Relevant documents with their distances.
         """
-
+        if (n_neighbours > len(self._db.get_vectors()) or not isinstance(query, str)
+                or query is None or len(query) == 0 or not isinstance(n_neighbours, int)):
+            raise ValueError
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/main.py]: .

File: lab_4_retrieval_w_clustering/settings.json [hunk 12]
```diff
@@ -1,3 +1,3 @@
 {
-    "target_score": 0
+    "target_score": 6
```
Comment (Vasilisa-Blyudova) [inline lab_4_retrieval_w_clustering/settings.json:2]: What about 8?

**Comment by Vasilisa-Blyudova** [comment]
No changes since last review

File: lab_4_retrieval_w_clustering/main.py (hunk 9 above)
Comment (CloudWhisp) [reply lab_4_retrieval_w_clustering/main.py]: What do you mean by "rm"? Remove, rename?

File: lab_4_retrieval_w_clustering/main.py (hunk 1 above)
Comment (CloudWhisp) [reply lab_4_retrieval_w_clustering/main.py:7]: fixed

File: lab_4_retrieval_w_clustering/main.py (hunk 2 above)
Comment (CloudWhisp) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 3 above)
Comment (CloudWhisp) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 4 above)
Comment (CloudWhisp) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 5 above)
Comment (CloudWhisp) [reply lab_4_retrieval_w_clustering/main.py]: fixed

File: lab_4_retrieval_w_clustering/main.py (hunk 6 above)
Comment (CloudWhisp) [reply lab_4_retrieval_w_clustering/main.py]: fixed

File: lab_4_retrieval_w_clustering/main.py (hunk 7 above)
Comment (CloudWhisp) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 8 above)
Comment (CloudWhisp) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 10 above)
Comment (CloudWhisp) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 11 above)
Comment (CloudWhisp) [reply lab_4_retrieval_w_clustering/main.py]: Done

**Pull request closed without merging**